File: src/openrgb_flowers/effects/flower_instance.py
"""Individual blooming flower model with organic lifecycle."""
from __future__ import annotations
from typing import Tuple
import numpy as np
from openrgb_flowers.core.interfaces.i_flower_model import IFlowerModel
from openrgb_flowers.core.interfaces.i_color_palette import IColorPalette
from openrgb_flowers.effects.petal_geometry import PetalGeometry
from openrgb_flowers.math.spatial_grid import SpatialGrid
from openrgb_flowers.math.easing import EasingFunctions


class FlowerInstance(IFlowerModel):
    """Encapsulates the lifecycle, petal geometry, and color propagation of a single blooming flower."""

    def __init__(
        self,
        center_x: float,
        center_y: float,
        palette: IColorPalette,
        petal_count: int = 5,
        max_radius: float = 0.45,
        lifespan: float = 3.2,
        rotation: float = 0.0,
        petal_depth: float = 0.22,
        spatial_grid: SpatialGrid | None = None,
    ) -> None:
        self._center_x = float(center_x)
        self._center_y = float(center_y)
        self._palette = palette
        self._petal_count = int(petal_count)
        self._max_radius = float(max_radius)
        self._lifespan = max(0.5, float(lifespan))
        self._age = 0.0
        self._rotation = float(rotation)
        self._petal_depth = float(petal_depth)
        self._spatial_grid = spatial_grid or SpatialGrid()

    @property
    def age(self) -> float:
        return self._age

    @property
    def lifespan(self) -> float:
        return self._lifespan

    def update(self, dt: float) -> bool:
        """Advances flower age by dt seconds. Returns True if alive."""
        self._age += dt
        return self.is_alive()

    def is_alive(self) -> bool:
        return self._age < self._lifespan

    def get_center(self) -> Tuple[float, float]:
        return (self._center_x, self._center_y)

    def get_palette(self) -> IColorPalette:
        return self._palette

    def get_lifecycle_phase(self) -> float:
        """Returns normalized lifecycle progress in [0.0, 1.0]."""
        return min(1.0, self._age / self._lifespan)

    def get_current_radius(self) -> float:
        """Calculates current physical expansion radius based on growth curve."""
        progress = self.get_lifecycle_phase()
        # Fast expansion in opening phase (0.0 -> 0.6), slowing down at full bloom
        growth_t = min(1.0, progress / 0.65)
        expansion = float(EasingFunctions.ease_out_cubic(growth_t))
        return self._max_radius * expansion

    def get_intensity_envelope(self) -> float:
        """Calculates temporal brightness envelope (budding -> peak bloom -> gentle wilt)."""
        p = self.get_lifecycle_phase()
        if p < 0.2:
            # Budding: smooth fade in
            return float(EasingFunctions.ease_in_out_cubic(p / 0.2))
        elif p < 0.65:
            # Peak bloom
            return 1.0
        else:
            # Wilting & dissolving: soft decay
            fade_t = (p - 0.65) / 0.35
            return float(1.0 - EasingFunctions.ease_in_out_cubic(fade_t))

    def evaluate(self, x_coords: np.ndarray, y_coords: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
        """Evaluates intensities and RGB colors across all keyboard key coordinates."""
        current_radius = self.get_current_radius()
        envelope = self.get_intensity_envelope()

        if current_radius < 0.01 or envelope < 0.01:
            n = len(x_coords)
            return np.zeros(n, dtype=np.float32), np.zeros((n, 3), dtype=np.float32)

        # Aspect-ratio corrected distances and angles
        _, _, distances, angles = self._spatial_grid.correct_aspect_ratio(
            x_coords, y_coords, self._center_x, self._center_y
        )

        # Calculate petal boundary radii
        boundary_radii = PetalGeometry.compute_petal_radii(
            angles, current_radius, self._petal_count, self._petal_depth, self._rotation
        )

        # Calculate spatial falloff
        falloff = PetalGeometry.compute_falloff(distances, boundary_radii, edge_softness=0.3)
        intensities = falloff * envelope

        # Sample colors along radial gradient: center (0.0) -> petal body -> tip (1.0)
        grad_pos = np.clip(distances / (boundary_radii + 1e-6), 0.0, 1.0)
        colors = self._palette.sample_vectorized(grad_pos)

        return intensities.astype(np.float32), colors.astype(np.float32)

File: src/effects/flower_instance.cpp
#include "openrgb_flowers/effects/flower_instance.hpp"
#include "openrgb_flowers/effects/petal_geometry.hpp"
#include "openrgb_flowers/math/easing.hpp"

#include <algorithm>
#include <cmath>

namespace openrgb_flowers::effects {

FlowerInstance::FlowerInstance(
    float center_x,
    float center_y,
    std::shared_ptr<core::interfaces::IColorPalette> palette,
    int petal_count,
    float max_radius,
    float lifespan,
    float rotation,
    float petal_depth,
    math::SpatialGrid grid
) : center_x_(center_x),
    center_y_(center_y),
    palette_(std::move(palette)),
    petal_count_(petal_count),
    max_radius_(max_radius),
    lifespan_(std::max(0.5f, lifespan)),
    age_(0.0f),
    rotation_(rotation),
    petal_depth_(petal_depth),
    grid_(grid) {}

bool FlowerInstance::update(float dt) {
    age_ += dt;
    return is_alive();
}

bool FlowerInstance::is_alive() const {
    return age_ < lifespan_;
}

std::pair<float, float> FlowerInstance::get_center() const {
    return {center_x_, center_y_};
}

float FlowerInstance::get_lifecycle_phase() const {
    return std::min(1.0f, age_ / lifespan_);
}

float FlowerInstance::get_current_radius() const {
    float progress = get_lifecycle_phase();
    float growth_t = std::min(1.0f, progress / 0.65f);
    float expansion = math::EasingFunctions::ease_out_cubic(growth_t);
    return max_radius_ * expansion;
}

float FlowerInstance::get_intensity_envelope() const {
    float p = get_lifecycle_phase();
    if (p < 0.2f) {
        return math::EasingFunctions::ease_in_out_cubic(p / 0.2f);
    } else if (p < 0.65f) {
        return 1.0f;
    } else {
        float fade_t = (p - 0.65f) / 0.35f;
        return 1.0f - math::EasingFunctions::ease_in_out_cubic(fade_t);
    }
}

void FlowerInstance::evaluate(
    const float* x_coords,
    const float* y_coords,
    float* out_intensities,
    core::models::ColorRGB* out_colors,
    size_t count
) const {
    float current_radius = get_current_radius();
    float envelope = get_intensity_envelope();

    if (current_radius < 0.01f || envelope < 0.01f) {
        for (size_t i = 0; i < count; ++i) {
            out_intensities[i] = 0.0f;
            out_colors[i] = core::models::ColorRGB(0, 0, 0);
        }
        return;
    }

    for (size_t i = 0; i < count; ++i) {
        auto [dx, dy, dist, angle] = grid_.correct_aspect_ratio_single(
            x_coords[i], y_coords[i], center_x_, center_y_
        );

        float boundary_radius = PetalGeometry::compute_petal_radius(
            angle, current_radius, petal_count_, petal_depth_, rotation_
        );

        float falloff = PetalGeometry::compute_falloff(dist, boundary_radius, 0.3f);
        float intensity = falloff * envelope;
        out_intensities[i] = intensity;

        float grad_pos = std::clamp(dist / (boundary_radius + 1e-6f), 0.0f, 1.0f);
        out_colors[i] = palette_->sample(grad_pos);
    }
}

} // namespace openrgb_flowers::effects
